Crash when destroying FrameView.

megangardner opened the pull request:
#### 67a39ee3032af6570dfd4e41293b3576fceeab6c
<pre>
Crash when destroying FrameView.
<a href="https://bugs.webkit.org/show_bug.cgi?id=268316">https://bugs.webkit.org/show_bug.cgi?id=268316</a>
<a href="https://rdar.apple.com/121636013">rdar://121636013</a>

Reviewed by Wenson Hsieh.

Based on the crash log and the crashlog script, we are missing
a null check.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::setWasScrolledByUser):

Canonical link: <a href="https://commits.webkit.org/273792@main">https://commits.webkit.org/273792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80eef9042369a249c68220582c0c9644c21b7260

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39320 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37901 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18079 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12723 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37224 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13155 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32421 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11512 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11528 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32683 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40567 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33219 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/33061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37452 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11793 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9625 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35569 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4750 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12680 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4677,8 +4677,9 @@ void LocalFrameView::setWasScrolledByUser(bool wasScrolledByUser)
 
-    if (wasScrolledByUser)
-        m_frame->document()->setGotoAnchorNeededAfterStylesheetsLoad(false);
+    auto document = m_frame->document();
```
```suggestion
    auto document = m_frame->protectedDocument();
```

whsieh replied:
Or alternately:
```suggestion
    RefPtr document = m_frame->document();
```
(not sure which one we prefer…)

cdumez replied:
We prefer:
```
RefPtr document = m_frame->document();
```

The `protectedFoo()` are used when there is no local variable.

megangardner committed:
Crash when destroying FrameView.
https://bugs.webkit.org/show_bug.cgi?id=268316
rdar://121636013

Reviewed by Wenson Hsieh.

Based on the crash log and the crashlog script, we are missing
a null check.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::setWasScrolledByUser):

Canonical link: https://commits.webkit.org/273792@main
